perfkitbenchmarker/scripts/object_storage_api_tests.py

```diff
@@ -518,6 +592,178 @@ def SingleStreamThroughputBenchmark(storage_schema, host_to_connect=None):
+      for start_time, latency in zip(res['start_times'], res['latencies']):
+        json.dump({'kind': res['kind'],
+                   'start_time': start_time,
+                   'latency': latency,
+                   'size': FLAGS.object_size_kb,
+                   'stream_num': stream_num},
+                  sys.stdout)
+        sys.stdout.write('\n')
+  except Queue.Empty:
+    pass
+
+  if num_results != FLAGS.num_streams:
+    raise Exception('Couldn\'t get results from all streams!')
+
+
+def MultiStreamWrites(storage_schema, host_to_connect=None):
+  """Run multi-stream write threads."""
+  RunMultiThreadedWorkersWriteResults(
+      WriteWorker,
+      (storage_schema,
+       host_to_connect,
+       FLAGS.object_size_kb,
+       FLAGS.objects_per_stream,
+       FLAGS.start_time),
+      storage_schema,
+      host_to_connect=host_to_connect)
+
+
+def MultiStreamReads(storage_schema, host_to_connect=None):
+  """Run multi-stream read threads."""
+  RunMultiThreadedWorkersWriteResults(
+      ReadWorker,
+      (storage_schema,
+       host_to_connect,
+       FLAGS.object_size_kb,
+       FLAGS.start_time),
+      storage_schema,
+      host_to_connect=host_to_connect)
+
+
+def SleepUntilTime(when, too_late=None):
+  """Sleep until a given time.
+
+  Args:
+    when: float. The time to sleep to, as a POSIX timestamp.
+
+    too_late: None or Exception. If when is before now + MIN_SLEEP,
+      raise the given exception. If too_late is None, log a message
+      but do not raise an exception.
+  """
+
+  now = time.time()
+  sleep_time = when - now
+  if sleep_time > MIN_SLEEP:
+    time.sleep(sleep_time)
+  elif too_late is not None:
+    raise too_late('Sleep time %s was too small' % sleep_time)
+  else:
+    logging.info('Sleep time %s was too small', sleep_time)
+
+
+OBJECTS_WRITTEN_FILE_PREFIX = 'pkb-objects-written-'
+
+
+def WriteWorker(storage_schema, host_to_connect,
+                object_size_kb, num_objects,
+                start_time, result_queue, worker_num):
+  payload_bytes, payload_string = GenerateWritePayload(
+      object_size_kb * BYTES_PER_KILOBYTE)
+
+  objects_written = []
+  start_times = []
+  latencies = []
+
+  if start_time is not None:
+    SleepUntilTime(start_time, too_late=Exception)
+
+  WriteObjectsFromBuffer(storage_schema, FLAGS.bucket,
+                         ('pkb_write_worker_%s' % worker_num),
+                         num_objects, payload_bytes, payload_string,
+                         object_size_kb * BYTES_PER_KILOBYTE,
+                         objects_written,
+                         object_start_times=start_times,
+                         latency_results=latencies,
+                         host_to_connect=host_to_connect)
+
+  objects_filename = os.path.join(TMP_DIR,
+                                  OBJECTS_WRITTEN_FILE_PREFIX + str(worker_num))
+
+  # Error handling deliberately omitted - abort on any exception.
+  if os.path.exists(objects_filename):
```

Comment:
A couple of additional questions regarding the object name files under /tmp:
1. Who is going to cleanup these files? Do we need to add logic in CleanupBucket() function?

2.I think we should remove all existing object state files under the /tmp directory, regardless of the worker number suffix. The use case we support is write once with X streams, and read it many times with Y1, Y2, ... Yn streams. If you try to do write tests more than once, the last write test will be retained and used for subsequent read test. 

This is simple and deterministic.